backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/JamesPostgresConnectionFactory.java

```diff
@@ -0,0 +1,37 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.postgres.utils;
+
+import java.util.Optional;
+
+import org.apache.james.core.Domain;
+
+import io.r2dbc.spi.Connection;
+import reactor.core.publisher.Mono;
+
+public interface JamesPostgresConnectionFactory {
+    String DOMAIN_ATTRIBUTE = "app.current_domain";
+
+    default Mono<? extends Connection> getConnection(Domain domain) {
+        return getConnection(Optional.ofNullable(domain));
+    }
+
+    Mono<? extends Connection> getConnection(Optional<Domain> domain);
```

Comment:
Q: Is `Mono<Connection> getConnection(Optional<Domain> domain);` enough?

Reply:
e.g. `new PostgresExecutor(postgresConnectionFactory.getConnection(Optional.empty()))`

-> compiler error: `PostgresExecutor(reactor.core.publisher.Mono<io.r2dbc.spi.Connection>)' in 'org.apache.james.backends.postgres.utils.PostgresExecutor' cannot be applied to '(reactor.core.publisher.Mono<capture<? extends io.r2dbc.spi.Connection>>)`

Then always casting Connection children types to Connection parent type is painful.

Reply:
changed all to Connection